br/pkg/lightning/backend/external/byte_reader.go

```diff
@@ -0,0 +1,162 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package external
+
+import (
+	"context"
+	"io"
+
+	"github.com/pingcap/tidb/br/pkg/storage"
+	"github.com/pingcap/tidb/util/logutil"
+	"github.com/pingcap/tidb/util/mathutil"
+	"go.uber.org/zap"
+)
+
+type byteReader struct {
+	ctx           context.Context
+	storageReader storage.ReadSeekCloser
+
+	buf       []byte
+	bufOffset int
+
+	auxBuf   []byte
+	cowSlice [][]byte // copy-on-write slices
+	isEOF    bool
+}
+
+func openStoreReaderAndSeek(
+	ctx context.Context,
+	store storage.ExternalStorage,
+	name string,
+	initFileOffset uint64,
+) (storage.ReadSeekCloser, error) {
+	storageReader, err := store.Open(ctx, name)
+	if err != nil {
+		return nil, err
+	}
+	_, err = storageReader.Seek(int64(initFileOffset), io.SeekStart)
+	if err != nil {
+		return nil, err
+	}
+	return storageReader, nil
+}
+
+func newByteReader(ctx context.Context, storageReader storage.ReadSeekCloser, bufSize int) *byteReader {
+	return &byteReader{
+		ctx:           ctx,
+		storageReader: storageReader,
+		buf:           make([]byte, bufSize),
+		bufOffset:     0,
+	}
+}
+
+type slice interface {
+	get() []byte
+}
+
+type commonSlice []byte
+
+func (c commonSlice) get() []byte {
+	return c
+}
+
+type lazySlice struct {
+	cowIdx     int
+	byteReader *byteReader
+}
+
+func (s lazySlice) get() []byte {
+	return s.byteReader.cowSlice[s.cowIdx]
+}
+
+// sliceNext reads the next n bytes from the reader and returns a buffer slice containing those bytes.
+func (r *byteReader) sliceNext(n int) (slice, error) {
+	b := r.next(n)
+	readLen := len(b)
+	if readLen == n {
+		return mkLazySlice(r, b), nil
+	}
+	// If the reader has fewer than n bytes remaining in current buffer,
+	// `auxBuf` is used as a container instead.
+	if cap(r.auxBuf) < n {
+		r.auxBuf = make([]byte, n)
+	}
+	r.auxBuf = r.auxBuf[:n]
+	copy(r.auxBuf, b)
+	for readLen < n {
+		r.cloneSlices()
+		err := r.reload()
+		if err != nil {
+			return nil, err
+		}
+		b = r.next(n - readLen)
+		copy(r.auxBuf[readLen:], b)
+		readLen += len(b)
+	}
+	return commonSlice(r.auxBuf), nil
+}
+
+func mkLazySlice(r *byteReader, s []byte) lazySlice {
+	r.cowSlice = append(r.cowSlice, s)
+	return lazySlice{
+		cowIdx:     len(r.cowSlice) - 1,
+		byteReader: r,
+	}
+}
+
+func (r *byteReader) reset() {
```

Comment:
How often will the caller invoke `reset`? Because I see this is the only method that reduce the `cowSlice`, if it's not very often called, I'm a bot worried that `sliceNext` will copy all the `cowSlice` at last so we can't avoid the copy overhead.

Reply:
And if the caller always call `reset` after every `sliceNext`, we don't need to implement `lazySlice`. The usage pattern is important I think

Reply:
OK I found the usage in #45356

Reply:
In https://github.com/pingcap/tidb/blob/1b723a219b1fbb5888b68f2ab53e010655a2534c/br/pkg/lightning/backend/sharedisk/reader.go#L52
we reset the cowSlice, but the slice is still used in the kvHeap
https://github.com/pingcap/tidb/blob/1b723a219b1fbb5888b68f2ab53e010655a2534c/br/pkg/lightning/backend/sharedisk/iter.go#L110-L118
So following sliceNext will overwrite the underlying array, and thus change the slice in kvHeap

```
func TestNextKV(t *testing.T) {
	ms := &mockExtStore{src: []byte("0123456789")}
	br := newByteReader(context.Background(), ms, 1)
	err := br.reload()
	require.NoError(t, err)

	nextKV := func() ([]byte, []byte) {
		br.reset()
		key, err := br.sliceNext(1)
		require.NoError(t, err)
		val, err := br.sliceNext(1)
		require.NoError(t, err)
		return key.get(), val.get()
	}
	k1, v1 := nextKV()
	require.Equal(t, []byte("0"), k1)
	require.Equal(t, []byte("1"), v1)
	k2, v2 := nextKV()
	require.Equal(t, []byte("0"), k1)
	require.Equal(t, []byte("1"), v1) // fail because underlying array is changed to "3"
	require.Equal(t, []byte("2"), k2)
	require.Equal(t, []byte("3"), v2)
}
```

Reply:
> we reset the cowSlice, but the slice is still used in the kvHeap.

Yes, you are correct. We need to guarantee that the slice is not used anymore before calling `nextKV()/reset()`.